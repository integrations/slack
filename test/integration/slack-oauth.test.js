const supertest = require('supertest');
const nock = require('nock');
const queryString = require('query-string');

const helper = require('.');

const { probot } = helper;

const fixtures = require('../fixtures');

const access = fixtures.slack.oauth.token;

const request = supertest.agent(probot.server);

describe('Integration: slack authentication', () => {
  beforeEach(() => {
    delete process.env.ALLOWED_TEAMS;
  });

  test('/login', async () => {
    const res = await request.get('/slack/oauth/login')
      .expect(302);

    const { location } = res.headers;
    const pattern = /^https:\/\/slack\.com\/oauth\/authorize\?client_id=(?:.*)&scope=(?:.*)&state=(?:.*)$/;
    expect(location).toMatch(pattern);

    const code = 'code-from-slack';
    const { state } = queryString.parse(location.replace('https://slack.com/oauth/authorize', ''));

    nock('https://slack.com').post('/api/oauth.token', {
      client_id: process.env.SLACK_CLIENT_ID,
      client_secret: process.env.SLACK_CLIENT_SECRET,
      code,
    }).reply(200, access);

    await request.get('/slack/oauth/callback').query({ code, state })
      .expect(302)
      .expect('Location', `https://slack.com/app_redirect?app=${access.app_id}&team=${access.team_id}`);

    const { SlackWorkspace } = helper.robot.models;
    const workspace = await SlackWorkspace.findOne({ where: { slackId: access.team_id } });
    expect(workspace.accessToken).toEqual(access.access_token);
  });

  test('updates the access token', async () => {
    const { SlackWorkspace } = helper.robot.models;

    const workspace = await SlackWorkspace.create({ slackId: access.team_id, accessToken: 'old' });

    const res = await request.get('/slack/oauth/login')
      .expect(302);
    const { location } = res.headers;
    const { state } = queryString.parse(location.replace('https://slack.com/oauth/authorize', ''));
    const code = 'code-from-slack';

    nock('https://slack.com').post('/api/oauth.token').reply(200, access);

    await request.get('/slack/oauth/callback')
      .query({ code, state })
      .expect(302);

    await workspace.reload();

    expect(workspace.accessToken).not.toEqual('old');
    expect(workspace.accessToken).toEqual(access.access_token);
  });

  test('denies teams that are not allowed', async () => {
    process.env.ALLOWED_TEAMS = 'not-your-team,not-your-team-either';

    nock('https://slack.com').post('/api/oauth.token').reply(200, access);
    nock('https://slack.com').post('/api/team.info').reply(200, fixtures.slack.team.info);
    nock('https://slack.com').post('/api/auth.revoke').reply(200, { ok: true });

    const res = await request.get('/slack/oauth/login')
      .expect(302);
    const { location } = res.headers;
    const { state } = queryString.parse(location.replace('https://slack.com/oauth/authorize', ''));
    const code = 'code-from-slack';

    await request.get('/slack/oauth/callback')
      .query({ code, state })
      .expect(302)
      .expect('Location', '/denied');
  });

  test('allows specified teams', async () => {
    process.env.ALLOWED_TEAMS = 'someone-else,example';

    nock('https://slack.com').post('/api/oauth.token').reply(200, access);
    nock('https://slack.com').post('/api/team.info').reply(200, fixtures.slack.team.info);

    const res = await request.get('/slack/oauth/login')
      .expect(302);
    const { location } = res.headers;
    const { state } = queryString.parse(location.replace('https://slack.com/oauth/authorize', ''));
    const code = 'code-from-slack';

    await request.get('/slack/oauth/callback')
      .query({ code, state })
      .expect(302)
      .expect('Location', `https://slack.com/app_redirect?app=${access.app_id}&team=${access.team_id}`);
  });

  test('user aborting oauth process redirects to restart OAuth flow', async () => {
    await request.get('/slack/oauth/callback').query({ error: 'access_denied' })
      .expect(302)
      .expect('Location', '/slack/oauth/login');
  });

  test('no state param redirects to restart OAuth flow', async () => {
    await request.get('/slack/oauth/callback').query({ code: 'abc' })
      .expect(302)
      .expect('Location', '/slack/oauth/login');
  });

  test('no session redirects to restart OAuth flow', async () => {
    await supertest(probot.server).get('/slack/oauth/callback').query({ code: 'abc', state: 'def' })
      .expect(302)
      .expect('Location', '/slack/oauth/login');
  });

  test('invalid state redirects to restart OAuth flow', async () => {
    await request.get('/slack/oauth/login')
      .expect(302);
    const code = 'code-from-slack';
    await request.get('/slack/oauth/callback').query({ code, state: 'not-the-same' })
      .expect(302)
      .expect('Location', '/slack/oauth/login');
  });

  test('slack non-ok response redirects to restart OAuth flow', async () => {
    const res = await request.get('/slack/oauth/login')
      .expect(302);
    const { location } = res.headers;
    const { state } = queryString.parse(location.replace('https://slack.com/oauth/authorize', ''));
    const code = 'code-from-slack';

    nock('https://slack.com').post('/api/oauth.token').reply(200, { ok: false, error: 'test_error' });

    await request.get('/slack/oauth/callback').query({ code, state })
      .expect(302)
      .expect('Location', '/slack/oauth/login');
  });
});
