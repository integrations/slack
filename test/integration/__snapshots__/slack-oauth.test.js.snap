// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: slack authentication /login 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"text\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\",\\"fallback\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\"},{\\"color\\":\\"#24292f\\",\\"text\\":\\"Looking for additional help? Try /github help\\",\\"fallback\\":\\"Looking for additional help? Try /github help\\"}]",
  "token": "xoxa-a-b-c-d",
}
`;

exports[`Integration: slack authentication allows all teams 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"text\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\",\\"fallback\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\"},{\\"color\\":\\"#24292f\\",\\"text\\":\\"Looking for additional help? Try /github help\\",\\"fallback\\":\\"Looking for additional help? Try /github help\\"}]",
  "token": "xoxa-a-b-c-d",
}
`;

exports[`Integration: slack authentication updates the access token 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"text\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\",\\"fallback\\":\\"You've successfully installed GitHub on this Slack workspace :tada:\\\\nTo subscribe a channel to a repository use the following slash command:\\\\n/github subscribe owner/repository\\\\n\\"},{\\"color\\":\\"#24292f\\",\\"text\\":\\"Looking for additional help? Try /github help\\",\\"fallback\\":\\"Looking for additional help? Try /github help\\"}]",
  "token": "xoxa-a-b-c-d",
}
`;
